file.go

```diff
@@ -0,0 +1,259 @@
+	}
+	readStr := func(n int, data *string) bool {
+		buf := make([]byte, n)
+		err = binary.Read(r, binary.LittleEndian, buf)
+		if err == nil {
+			*data = cstring(buf)
+		}
+		return err == nil
+	}
+
+	// the Metadata header is defined in §II.24.2.1.
+	var hdr struct {
+		Signature    uint32
+		MajorVersion uint16
+		MinorVersion uint16
+		Reserved     uint32
+		Version      string
+		Flags        uint16
+	}
+	if !read(&hdr.Signature) {
+		return "", nil, fmt.Errorf("failure to read the metadata header signature: %v", err)
+	}
+	// magic signature from II.24.2.1
+	const signature = 0x424A5342
+	if hdr.Signature != signature {
+		return "", nil, fmt.Errorf("metadata header signature (%#X) must be (%#X)", hdr.Signature, signature)
+	}
+	var streamsCount uint16
+	var cstringLength uint32
+	if !read(&hdr.MajorVersion) ||
+		!read(&hdr.MinorVersion) ||
+		!read(&hdr.Reserved) ||
+		!read(&cstringLength) ||
+		!readStr(int(cstringLength), &hdr.Version) ||
+		!read(&hdr.Flags) ||
+		!read(&streamsCount) {
+		return "", nil, fmt.Errorf("failure to read the metadata header: %v", err)
+	}
+
+	readStreamNameStr := func(data *string) bool {
+		// the name of the stream as null-terminated variable length array
+		// of ASCII characters, padded to the next 4-byte boundary
+		// with \x00 characters. The name is limited to 32 characters.
+		const nameMaxLength = 32
+		const namePadding = 4
+		var nameBuf [nameMaxLength]byte
+		// Read in chunks of 4 bytes, accumulating the string
+		// into nameBuf until the first \x00 character is found.
+		var found bool
+		for j := 0; j < nameMaxLength; j += namePadding {
+			if !read(nameBuf[j : j+namePadding]) {
+				break
+			}
+			idx := bytes.IndexByte(nameBuf[:j+namePadding], 0)
+			if idx != -1 {
+				found = true
+				*data = string(nameBuf[:idx])
+				break
+			}
+		}
+		if !found {
+			err = errors.New("name not found")
+		}
+		return err == nil
+	}
+
+	// parse stream headers.
+	heaps := make([]*Heap, streamsCount)
+	for i := 0; i < int(streamsCount); i++ {
+		// the stream header is defined in §II.24.2.2.
+		var s struct {
+			Offset uint32
+			Size   uint32
+			Name   string
+		}
+		if !read(&s.Offset) ||
+			!read(&s.Size) ||
+			!readStreamNameStr(&s.Name) {
+			return "", nil, fmt.Errorf("failure to read the stream header (%d): %v", i, err)
+		}
+		sr := io.NewSectionReader(ds, rootOffset+int64(s.Offset), int64(s.Size))
+		heaps[i] = &Heap{
+			sr:       sr,
+			ReaderAt: sr,
+			name:     s.Name,
+			Size:     s.Size,
+		}
+	}
+	return hdr.Version, heaps, nil
+}
+
+// sectionByRVA returns the section which contains rva.
+func sectionByRVA(pefile *pe.File, rva uint32) *pe.Section {
+	var ds *pe.Section
+	for _, s := range pefile.Sections {
+		if s.VirtualAddress <= rva && rva < s.VirtualAddress+s.VirtualSize {
+			ds = s
+			break
+		}
+	}
+	return ds
+}
+
+// cstring converts UTF8 byte sequence b to string.
+// It stops once it finds 0 or reaches end of b.
+func cstring(b []byte) string {
```

Comment:
Do we need to handle `b == nil` here?

Reply:
`b` is always initialized, but as `cstring` is only used in one place now I'll move its body where it belongs so there are no doubts about it.